versions/development/SPEC.md

```diff
@@ -698,6 +699,51 @@ Pair values can also be specified within the [workflow inputs JSON](https://gith
 }
 ```
 
+### Optional literals
+
+Any non-optional value can be stored into an optional variable by assigning it to an optional variable.
+This does not change the value, but it changes the type of the value:
+
+```WDL
+Int? maybe_five = 5
+```
+
+The `None` literal is the value that an optional has when it is not defined.
+
+```WDL
+Int? maybe_five_but_is_not = None
```

Comment:
Curious what the non-scalawags think - is `None` intuitive here? 

It is to me, but I'm used to Scala's `Option`, which is what this construct was modeled after. 

I don't know that I have any suggestion which seems better (going back to "it is intuitive to me"), perhaps `Empty`?

Reply:
In python, there is also a [`None`](https://stackoverflow.com/questions/19473185/what-is-a-none-value). It also means something like 'empty' and calling a `None` variable, will give a not defined error. So for people who know python, it can be intuitive (it is for me,  but I also know scala and I can not judge for other people).

Reply:
I think `null` is a better choice as that's how you write this concept in JSON, which is how WDL inputs are represented. That way the same concept looks the same in both paces.

Reply:
I think assigning `null` to something is a contradiction in itself. So, in my preference `null` is not available as a variable in the toolbox.

As for the "then it will look the same as in JSON" argument. You have a point there, but frankly, `None` is different from `null`. Using the same name for a different concept does not sound like a good idea to me.

Reply:
> One thing worth thinking through is if there's a meaningful difference between "This space intentionally left blank" and "For unknown reasons there's nothing here" in WDL-land. When those are both viable states I think that null should always mean the latter, and only the latter

Can you explain how these two different states could arise in WDL? They sound to me like a states in the mind of the WDL user, not the WDL interpreter.

Reply:
> I think assigning null to something is a contradiction in itself. So, in my preference null is not available as a variable in the toolbox.

Remember that, in my proposal, `null` is the name WDL has for the concept of an unassigned value. It has no meaning with respect to the implementation of WDL, or pointers, or other programming language concepts.

> As for the "then it will look the same as in JSON" argument. You have a point there, but frankly, None is different from null. Using the same name for a different concept does not sound like a good idea to me.

Are you saying that using `null` in JSON has a different effect than using (as you propose) `None` in WDL? If so, how is this difference detectable from inside WDL?

In my proposal, `null` is not different than `None` as they are just different spellings of the same concept. Maybe the point that you're trying to make is that `null` has baggage due to its meaning in other programming languages. I agree that it could be a hurdle for WDL users that have experience programming in other languages. Is this the argument that you're making here?

Reply:
@pshapiro4broad 

> Can you explain how these two different states could arise in WDL?

I don't have an example, just thought it was worth throwing it out there in case it prompted someone to think of one 

> In my proposal, null is not different than None as they are just different spellings of the same concept.

I don't think they necessarily are, going back to the difference I cited above. For instance in R `NA` and `NULL` are two very difference concepts, but they'd both fold into `null` in many languages which only have the one concept.

> Maybe the point that you're trying to make is that null has baggage due to its meaning in other programming languages. 

I certainly fall under that category (I've protested against adding "the billion dollar mistake" to WDL in the past as well), but to go back earlier in the convo I also admit I'm a fairly biased witness here.

Reply:
@geoffjentry @pshapiro4broad Most of the "Null vs None" stuff was already discussed in #209. Most people liked `None` more, so I settled for that.